common/notification/index.ts

```diff
@@ -456,29 +454,102 @@ export async function actionTaken<ID extends ActionID>(
                           OR: [{ contextID: null }, { contextID: notificationContext.uniqueId }],
                       }
                     : {}),
-            },
-        });
+            };
+
+            if (!dryRun) {
+                const dismissed = await prisma.concrete_notification.updateMany({
+                    data: {
+                        state: ConcreteNotificationState.ACTION_TAKEN,
+                        sentAt: new Date(now),
+                    },
+                    where: queryToDismiss,
+                });
+
+                logger.debug(`Notification.actionTaken dismissed ${dismissed.count} pending notifications`);
+                dismissedCount = dismissed.count;
+            } else {
+                dismissedCount = await prisma.concrete_notification.count({
+                    where: queryToDismiss,
+                });
+            }
+        }
 
-        logger.debug(`Notification.actionTaken dismissed ${dismissed.count} pending notifications`);
+        // --------------- Determine which Notifications to send directly and which to schedule ---------
 
-        const reminders = relevantNotifications.toSend.filter((it) => it.delay);
-        const directSends = relevantNotifications.toSend.filter((it) => !it.delay);
+        for (const notification of relevantNotifications.toSend) {
+            if (noDuplicates) {
+                assert(notificationContext.uniqueId, 'If noDuplicates is set, a uniqueId shall be set');
 
-        logger.debug(`Notification.actionTaken found reminders ${reminders.map((it) => it.id)} and directSends ${directSends.map((it) => it.id)}`);
+                const existingNotifications = await prisma.concrete_notification.count({
```

Comment:
The name `existingNotifications` is misleading imho. Maybe `existingDuplicates` or `sameNotifications`?

Reply:
or we cast it to a boolean and call it `hasExistingDuplicates`.